web/ext/csrf.py

```diff
@@ -0,0 +1,116 @@
+# encoding: utf-8
+
+import hashlib, hmac
+
+from binascii import hexlify, unhexlify
+from os import urandom
+
+from web.core.util import lazy
+from web.security.csrf import CSRFToken, SignedCSRFToken, CSRFError
+
+
+log = __import__('logging').getLogger(__name__)
+
+
+class CSRFHandler(object):
+    __slots__ = ('_ctx', 'ext', '_user_secret')
+    
+    def __init__(self, context, csrf_ext):
+        self._ctx = context
+        self.ext = csrf_ext
+        self._user_secret = None
+    
+    def __call__(self, token):
+        if len(token) != 128:
+            return False
+        
+        try:
+            SignedCSRFToken(token, secret=str(self.user_secret))
+        except CSRFError:
+            return False
+        return True
+    
+    def _generate(self):
+        return SignedCSRFToken(secret=str(self.user_secret))
+    
+    def __str__(self):
+        return self._generate().signed.decode('utf-8')
+        
+    def __bytes__(self):
+        return unhexlify(self._generate().signed)
+    
+    @property
+    def user_secret(self):
+        if self._user_secret:
+            return self._user_secret
+        
+        context = self._ctx
+        token = None
+        try:
+            if self.ext.cookie:
+                cookie = context.request.cookies.get(self.ext.cookie['name'], None)
+                if cookie: token = SignedCSRFToken(cookie, secret=self.ext._secret)
+            else:
+                token = context.session.csrf
```

Comment:
Right, it might also be an instance of CSRFToken in the case of using MemorySession, so it'll have to check for the existence of a value attr and use that if it exists.